File: src/game.c
#include "game.h"
#include <stdlib.h>

static int check_square_for_color(Game *g, char C);

void game_init(Game *g) {
    g->cells = NULL;
    g->started = 0;
    g->N = 0;
    g->moves_count = 0;
}

void game_free(Game *g) {
    if (g->cells) free(g->cells);
    g->cells = NULL;
}

int game_start(Game *g, int N, Player p1, Player p2) {
    if (N <= 2) return 0;
    game_free(g);
    g->N = N;
    g->cells = malloc(N*N);
    for (int i = 0; i < N*N; i++) g->cells[i] = '.';
    g->p1 = p1;
    g->p2 = p2;
    g->cur = 1;
    g->started = 1;
    g->moves_count = 0;
    return 1;
}

int game_in_bounds(Game *g, int x, int y) {
    return x >= 0 && y >= 0 && x < g->N && y < g->N;
}

char game_get_cell(Game *g, int x, int y) {
    return g->cells[y * g->N + x];
}

void game_set_cell(Game *g, int x, int y, char c) {
    g->cells[y * g->N + x] = c;
}

Player *game_cur_player(Game *g) {
    return (g->cur == 1) ? &g->p1 : &g->p2;
}

Player *game_other_player(Game *g) {
    return (g->cur == 1) ? &g->p2 : &g->p1;
}

void game_switch_turn(Game *g) {
    g->cur = (g->cur == 1) ? 2 : 1;
}

int game_board_full(Game *g) {
    return g->moves_count >= g->N * g->N;
}

int game_check_winner(Game *g, char *out_color) {
    if (check_square_for_color(g, 'W')) { *out_color = 'W'; return 1; }
    if (check_square_for_color(g, 'B')) { *out_color = 'B'; return 1; }
    return 0;
}

/* --- Проверка квадрата (любой ориентации) --- */
static int check_square_for_color(Game *g, char C) {
    int N = g->N;
    for (int ay = 0; ay < N; ay++) {
        for (int ax = 0; ax < N; ax++) {
            if (game_get_cell(g, ax, ay) != C) continue;
            for (int by = ay; by < N; by++) {
                for (int bx = (by == ay ? ax + 1 : 0); bx < N; bx++) {
                    if (game_get_cell(g, bx, by) != C) continue;

                    int dx = bx - ax;
                    int dy = by - ay;

                    // поворот влево
                    int cx = ax - dy, cy = ay + dx;
                    int dx2 = bx - dy, dy2 = by + dx;
                    if (game_in_bounds(g, cx, cy) &&
                        game_in_bounds(g, dx2, dy2) &&
                        game_get_cell(g, cx, cy) == C &&
                        game_get_cell(g, dx2, dy2) == C) {
                        return 1;
                    }

                    // поворот вправо
                    cx = ax + dy; cy = ay - dx;
                    dx2 = bx + dy; dy2 = by - dx;
                    if (game_in_bounds(g, cx, cy) &&
                        game_in_bounds(g, dx2, dy2) &&
                        game_get_cell(g, cx, cy) == C &&
                        game_get_cell(g, dx2, dy2) == C) {
                        return 1;
                    }
                }
            }
        }
    }
    return 0;
}


/* --- Логика компьютера (как раньше) --- */
static int would_win(Game *g,int x,int y,char C) {
    if (!game_in_bounds(g,x,y)) return 0;
    if (game_get_cell(g,x,y)!='.') return 0;
    game_set_cell(g,x,y,C); g->moves_count++;
    int ok=check_square_for_color(g,C);
    game_set_cell(g,x,y,'.'); g->moves_count--;
    return ok;
}
static int find_winning(Game *g,char C,int *rx,int *ry){
    for(int y=0;y<g->N;y++)for(int x=0;x<g->N;x++)
        if(game_get_cell(g,x,y)=='.'&&would_win(g,x,y,C)){*rx=x;*ry=y;return 1;}
    return 0;
}
void game_comp_move(Game *g,int *outx,int *outy){
    char C=game_cur_player(g)->color;
    char O=(C=='W'?'B':'W');
    int x,y;
    if(find_winning(g,C,&x,&y)){game_set_cell(g,x,y,C);g->moves_count++;*outx=x;*outy=y;return;}
    if(find_winning(g,O,&x,&y)){game_set_cell(g,x,y,C);g->moves_count++;*outx=x;*outy=y;return;}
    int freecount=0;
    for(int yy=0;yy<g->N;yy++)for(int xx=0;xx<g->N;xx++)if(game_get_cell(g,xx,yy)=='.')freecount++;
    int r=rand()%freecount,idx=0;
    for(int yy=0;yy<g->N;yy++)for(int xx=0;xx<g->N;xx++)if(game_get_cell(g,xx,yy)=='.'){if(idx==r){game_set_cell(g,xx,yy,C);g->moves_count++;*outx=xx;*outy=yy;return;}idx++;}
    *outx=*outy=-1;
}

File: game/src/main/java/org/task1/core/Game.java
package org.task1.core;

import java.util.Random;

public class Game {
    private char[] cells;
    private int N;
    private int movesCount;
    private int cur; // 1 или 2
    private boolean started;

    private Player p1;
    private Player p2;

    private final Random random = new Random();

    public void free() {
        cells = null;
    }

    public boolean start(int N, Player p1, Player p2) {
        if (N <= 2) return false;
        free();
        this.N = N;
        this.cells = new char[N * N];
        for (int i = 0; i < N * N; i++) {
            cells[i] = '.';
        }
        this.p1 = p1;
        this.p2 = p2;
        this.cur = 1;
        this.started = true;
        this.movesCount = 0;
        return true;
    }

    public boolean inBounds(int x, int y) {
        return x >= 0 && y >= 0 && x < N && y < N;
    }

    public char getCell(int x, int y) {
        return cells[y * N + x];
    }

    public void setCell(int x, int y, char c) {
        cells[y * N + x] = c;
    }

    public Player getCurPlayer() {
        return (cur == 1) ? p1 : p2;
    }

    public void switchTurn() {
        cur = (cur == 1) ? 2 : 1;
    }

    public boolean isBoardFull() {
        return movesCount >= N * N;
    }

    public char checkWinner() {
        if (checkSquareForColor('W')) return 'W';
        if (checkSquareForColor('B')) return 'B';
        return '.';
    }

    private boolean checkSquareForColor(char C) {
        for (int ay = 0; ay < N; ay++) {
            for (int ax = 0; ax < N; ax++) {
                if (getCell(ax, ay) != C) continue;
                for (int by = ay; by < N; by++) {
                    for (int bx = (by == ay ? ax + 1 : 0); bx < N; bx++) {
                        if (getCell(bx, by) != C) continue;

                        int dx = bx - ax;
                        int dy = by - ay;

                        int cx = ax - dy, cy = ay + dx;
                        int dx2 = bx - dy, dy2 = by + dx;
                        if (inBounds(cx, cy) && inBounds(dx2, dy2) &&
                                getCell(cx, cy) == C && getCell(dx2, dy2) == C) {
                            return true;
                        }

                        cx = ax + dy; cy = ay - dx;
                        dx2 = bx + dy; dy2 = by - dx;
                        if (inBounds(cx, cy) && inBounds(dx2, dy2) &&
                                getCell(cx, cy) == C && getCell(dx2, dy2) == C) {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }

    private boolean wouldWin(int x, int y, char C) {
        if (!inBounds(x, y)) return false;
        if (getCell(x, y) != '.') return false;
        setCell(x, y, C); movesCount++;
        boolean ok = checkSquareForColor(C);
        setCell(x, y, '.'); movesCount--;
        return ok;
    }

    private boolean findWinning(char C, int[] result) {
        for (int y = 0; y < N; y++) {
            for (int x = 0; x < N; x++) {
                if (getCell(x, y) == '.' && wouldWin(x, y, C)) {
                    result[0] = x;
                    result[1] = y;
                    return true;
                }
            }
        }
        return false;
    }

    public int[] compMove() {
        char C = getCurPlayer().getColor();
        char O = (C == 'W') ? 'B' : 'W';
        int[] res = new int[2];

        if (findWinning(C, res)) {
            setCell(res[0], res[1], C);
            movesCount++;
            return res;
        }
        if (findWinning(O, res)) {
            setCell(res[0], res[1], C);
            movesCount++;
            return res;
        }

        int freecount = 0;
        for (int yy = 0; yy < N; yy++) {
            for (int xx = 0; xx < N; xx++) {
                if (getCell(xx, yy) == '.') freecount++;
            }
        }
        int r = random.nextInt(freecount);
        int idx = 0;
        for (int yy = 0; yy < N; yy++) {
            for (int xx = 0; xx < N; xx++) {
                if (getCell(xx, yy) == '.') {
                    if (idx == r) {
                        setCell(xx, yy, C);
                        movesCount++;
                        res[0] = xx; res[1] = yy;
                        return res;
                    }
                    idx++;
                }
            }
        }
        return new int[]{-1, -1};
    }



    public void incMovesCount() { movesCount++; }

    public char[] getCells() {
        return cells;
    }

    public void setCells(char[] cells) {
        this.cells = cells;
    }

    public int getN() {
        return N;
    }

    public void setN(int n) {
        N = n;
    }

    public int getMovesCount() {
        return movesCount;
    }

    public void setMovesCount(int movesCount) {
        this.movesCount = movesCount;
    }

    public int getCur() {
        return cur;
    }

    public void setCur(int cur) {
        this.cur = cur;
    }

    public boolean isStarted() {
        return started;
    }

    public void setStarted(boolean started) {
        this.started = started;
    }

    public Player getP1() {
        return p1;
    }

    public void setP1(Player p1) {
        this.p1 = p1;
    }

    public Player getP2() {
        return p2;
    }

    public void setP2(Player p2) {
        this.p2 = p2;
    }
}
